proposals/crd-support.md

```diff
@@ -0,0 +1,52 @@
+
+## Goals
+* The solution should not be dependent on versioned CRDs. Even though these are in development, a solution needs to work with existing CRDs. But we would like a solution that can work with/take advantage of versioned CRDs when they become available.
+* It should support existing installed charts that use CRDs, regardless of which of the existing ways to install CRDs was used.
+* CRDs should never be deleted unless the user has explicitely asked for it. Since CRDs are global in a Kubernetes cluster, we can never know for sure that it is no longer in use.
+* CRDs should never be updated unless the user has explicitely asked for it. Updating CRDs could render them unusable for some of the services/applications that depend on them.
+* Mixing CRDs and CRs of the type in the same chart should be supported.
+
+
+## Implementation
+This proposal takes a very conservative approach on any changes to CRDs, but the hope is that it will not be to inconvenient. If CRDs are either only used by a single chart, or is used by many charts but rarely updated (like App CRD), updates should usually be possible without too much inconvenience for the user.
+
+CRDs need to be handled correctly through all the phases of a Helm release:
+
+### Install
+* CRDs should be the first resource processed during installation. CRDs must be installed into the cluster before any CRs belonging to the CRD.
+* If a chart contains one or more CRDs, the helm install process should for each of them, check if the CRD is already installed. 
+  * If it is, Helm should check if the installed CRD is identical to the one contained in the chart. 
+    * If it is, the install process can skip installing the CRD and just continue.
+    * If it is different, the install process should fail, unless the user has explicitly confirmed through a command line flag that the CRD in the chart should overwrite the existing one. If a CRD is versioned, it should allow multiple different versions of the CRD to coexist in the cluster. Overwrite will only apply if a chart has a CRD with a version that already exists.
+  * If it does not already exist, the CRD will be installed.
+* All CRDs in the chart will be tracked as part of a release when the chart is installed
+  * If the CRD was already installed, then the existing CRD will be adopted into the release.
+
+### Delete
+* When a chart is deleted, all CRDs will be orphaned, unless the user has explicitly through a command line flag decided that it should be deleted.
+ * CRs that were part of the release would be deleted with the release, even if the CRD is orphaned.
+
+### Update
+* Any new CRDs included in a new version of a chart will be handled in the same way that CRDs are handled in the install flow.
+* Any CRDs deleted in a new version of a chart will be handled in the same way that CRDs are handled in the delete flow.
```

Comment:
Deleting CRDs is complicated. For example, if another chart has a CR for the CRD and the CRD is deleted... the CRs in the other chart will delete. Doing the "update dance" will not be easy and people should have a deprecation period.

It would be nice, at least for a cluster admin, to see how many CRs are using a CRD.

Reply:
Yeah, deleting a CRD is very dangerous unless you have a good understanding of every possible use of the CRD. 
Looking up all CRs using a CRD is easy with just `kubectl get <name of crd>`, so maybe there can be some support in Helm to help users safely delete CRDs.